src/drivers/intel/ssp/ssp.c

```diff
@@ -699,6 +699,17 @@ static int ssp_set_config_tplg(struct dai *dai, struct ipc_config_dai *common_co
 	ssp->state[DAI_DIR_CAPTURE] = COMP_STATE_PREPARE;
 
 clk:
+	/* MCLK always-on: turn on mclk and never turn it off */
+	if (ssp->params.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_MCLK_AON) {
+		ret = ssp_mclk_prepare_enable(dai);
```

Comment:
@plbossart  I don't get your idea.   As I know,  ROM gets FW image in imr memory and doesn't need to do validation with CSME in IMR boot, but FW still needs to be set up by dsp.  Why ssp>clk_active never be released ?

Reply:
@RanderWang it's my understanding that the context is saved to IMR and restored from IMR, which means this ssp->clk_active value would be restored to its pre-suspend value.

@ujfalusi can you confirm if we store the image or the context?

Reply:
@plbossart, we store the firmware image for sure, I believe any dynamically allocated struct or data will be discarded.

The DSP and firmware needs to be re-initialized after IMR boot, but, but I have a recollection that in some cases the IMR booted firmware retained some information from it's previous life and that caused firmware crash (until we purged the IMR storage).

Hrm, let me think how to prove it either way...

Reply:
The machine driver would need to do a 'clk_enable' but for this we'd need to expose a clock provider and then that clock provided (an SOF client likely) would send the IPC. This is how it was done in the KBL time frame and it's the right way to deal with such requirements. The problem is that we don't have this IPC on old devices.

Reply:
> @plbossart, we store the firmware image for sure, I believe any dynamically allocated struct or data will be discarded.
> 
> The DSP and firmware needs to be re-initialized after IMR boot, but, but I have a recollection that in some cases the IMR booted firmware retained some information from it's previous life and that caused firmware crash (until we purged the IMR storage).
> 
> Hrm, let me think how to prove it either way...

I am pretty sure the IMR stuff relies on a context save of all the SRAM values into IMR, and restored.
If it was just a matter of re-executing the same firmware, we would not need any context save at all :-)

Reply:
> I am pretty sure the IMR stuff relies on a context save of all the SRAM values into IMR, and restored.
> If it was just a matter of re-executing the same firmware, we would not need any context save at all :-)

Not atm, no context is saved in IPC3. IMR usage in current code just saves the FW image from the code loader and boots directly from it rather than using the code loader (to copy to IMR for boot)